File: backend/app/schemas/execution_engine.py
from typing import Literal

from pydantic import BaseModel, Field


StepStatus = Literal["pending", "running", "completed", "failed"]
ThinkingStatus = Literal["pending", "running", "completed"]
PlanStatus = Literal["pending", "running", "completed", "failed"]
ReplayFrameType = Literal[
    "plan_created",
    "plan_running",
    "step_running",
    "step_completed",
    "step_failed",
    "plan_completed",
    "plan_failed",
]


class ExecutionEnquiryCreate(BaseModel):
    enquiry: str = Field(min_length=10)


class RiskAnalysisItem(BaseModel):
    risk: str
    impact: str
    mitigation: str


class ThinkingPanelStep(BaseModel):
    id: str
    label: str
    detail: str
    status: ThinkingStatus = "pending"


class DecisionItem(BaseModel):
    title: str
    why: str
    evidence: list[str] = Field(default_factory=list)
    selected_option: str
    confidence: str


class BusinessMemoryItem(BaseModel):
    execution_id: str | None = None
    title: str
    summary: str
    similarity_reason: str
    influence: str


class FinalRecommendation(BaseModel):
    supplier: str
    products: list[str] = Field(default_factory=list)
    estimated_quote: str
    expected_margin: str
    delivery_timeline: str
    approval_required: str
    business_reasoning: str


class WorkflowStep(BaseModel):
    id: str
    title: str
    description: str
    reasoning: str
    expected_output: str
    input_summary: str = ""
    output_summary: str = ""
    execution_time_seconds: float = 0.0
    assigned_agent: str
    order: int
    status: StepStatus = "pending"


class ExecutionSnapshot(BaseModel):
    id: str
    type: ReplayFrameType
    timestamp: str
    note: str
    active_step_id: str | None = None
    active_step_title: str | None = None
    workflow_state: list[WorkflowStep] = Field(default_factory=list)


class ExecutionPlanContent(BaseModel):
    intent: str
    required_tasks: list[str] = Field(default_factory=list)
    required_agents: list[str] = Field(default_factory=list)
    thinking_panel: list[ThinkingPanelStep] = Field(default_factory=list)
    decision_panel: list[DecisionItem] = Field(default_factory=list)
    workflow_steps: list[WorkflowStep] = Field(default_factory=list)
    risk_analysis: list[RiskAnalysisItem] = Field(default_factory=list)
    approval_requirement: str
    approval_status: str = "pending"
    final_recommendation: FinalRecommendation | None = None
    business_memory: list[BusinessMemoryItem] = Field(default_factory=list)
    execution_history: list[ExecutionSnapshot] = Field(default_factory=list)


class ApprovalDecisionRequest(BaseModel):
    decision: Literal["approved", "rejected", "changes_requested"]
    note: str | None = None


class RecommendationUpdateRequest(BaseModel):
    supplier: str | None = None
    products: list[str] | None = None
    expected_margin: str | None = None
    delivery_timeline: str | None = None
    estimated_quote: str | None = None


class ExecutionPlanRecord(BaseModel):
    id: str
    enquiry: str
    status: PlanStatus
    execution_plan: ExecutionPlanContent
    created_at: str | None = None
    updated_at: str | None = None

File: frontend/src/services/executionEngine.ts
import axios from "axios";

export type StepStatus = "pending" | "running" | "completed" | "failed";
export type PlanStatus = "pending" | "running" | "completed" | "failed";
export type ReplayFrameType =
  | "plan_created"
  | "plan_running"
  | "step_running"
  | "step_completed"
  | "step_failed"
  | "plan_completed"
  | "plan_failed";

export type ThinkingStatus = "pending" | "running" | "completed";

export type WorkflowStep = {
  id: string;
  title: string;
  description: string;
  reasoning: string;
  expected_output: string;
  input_summary: string;
  output_summary: string;
  execution_time_seconds: number;
  assigned_agent: string;
  order: number;
  status: StepStatus;
};

export type ThinkingPanelStep = {
  id: string;
  label: string;
  detail: string;
  status: ThinkingStatus;
};

export type DecisionItem = {
  title: string;
  why: string;
  evidence: string[];
  selected_option: string;
  confidence: string;
};

export type BusinessMemoryItem = {
  execution_id?: string | null;
  title: string;
  summary: string;
  similarity_reason: string;
  influence: string;
};

export type ExecutionSnapshot = {
  id: string;
  type: ReplayFrameType;
  timestamp: string;
  note: string;
  active_step_id?: string | null;
  active_step_title?: string | null;
  workflow_state: WorkflowStep[];
};

export type RiskAnalysisItem = {
  risk: string;
  impact: string;
  mitigation: string;
};

export type FinalRecommendation = {
  supplier: string;
  products: string[];
  estimated_quote: string;
  expected_margin: string;
  delivery_timeline: string;
  approval_required: string;
  business_reasoning: string;
  confidence_score?: string;
};

export type ExecutionPlanContent = {
  intent: string;
  required_tasks: string[];
  required_agents: string[];
  thinking_panel: ThinkingPanelStep[];
  decision_panel: DecisionItem[];
  workflow_steps: WorkflowStep[];
  risk_analysis: RiskAnalysisItem[];
  approval_requirement: string;
  approval_status: string;
  final_recommendation?: FinalRecommendation | null;
  business_memory: BusinessMemoryItem[];
  execution_history: ExecutionSnapshot[];
};

export type RecommendationUpdateRequest = {
  supplier?: string;
  products?: string[];
  expected_margin?: string;
  delivery_timeline?: string;
  estimated_quote?: string;
};

export type ExecutionPlanRecord = {
  id: string;
  enquiry: string;
  status: PlanStatus;
  execution_plan: ExecutionPlanContent;
  created_at?: string | null;
  updated_at?: string | null;
};

export type ExecutionPlanResponse = {
  message: string;
  data: ExecutionPlanRecord;
};

const api = axios.create({
  baseURL: import.meta.env.VITE_API_URL ?? "http://localhost:8000",
});

export async function createExecutionPlan(enquiry: string) {
  const response = await api.post<ExecutionPlanResponse>("/execution-engine/plans", {
    enquiry,
  });
  return response.data;
}

export async function getExecutionPlan(planId: string) {
  const response = await api.get<ExecutionPlanRecord>(`/execution-engine/plans/${planId}`);
  return response.data;
}

export async function getExecutionPlans() {
  const response = await api.get<ExecutionPlanRecord[]>("/execution-engine/plans");
  return response.data;
}

export async function approveExecutionPlan(planId: string, decision: "approved" | "rejected" | "changes_requested", note?: string) {
  const response = await api.post<ExecutionPlanResponse>(`/execution-engine/plans/${planId}/approval`, {
    decision,
    note,
  });
  return response.data;
}

export async function updateFinalRecommendation(planId: string, payload: RecommendationUpdateRequest) {
  const response = await api.patch<ExecutionPlanResponse>(`/execution-engine/plans/${planId}/recommendation`, payload);
  return response.data;
}
